### ISSUE TITLE
Stuck in GC loop even with high xmx

### ISSUE THREAD

**Issue by maxhillaert** [scalameter/scalameter#209]
Hi,

I've added cover with scalameter to an algorithm.   I then intentionally make it worse by doing a bad concat of lists in a tight loop.

Problem is that its so bad it never seems to fail and scalameter output keeps saying "GC detected " and sometimes  "Reinstating BenchMark Value"

The memory will have exploded to 4GB, but just sit there,, even though I set XMX 10000m.

This is how the JVM is loaded:

Starting new JVM: java -server -Xmx10048m -Xms4048m -cp C:\Users\hillmax\AppData\Local\Temp\classpath1003926506.jar;C:\Users\hillmax\AppData\Roaming\JetBrains\IntelliJ IDEA Community Edition 2018.1.1\lib\idea_rt.jar org.scalameter.execution.Main C:\Users\hillmax\AppData\Local\Temp\newjvm-5758005586671152155-io

Is XMX really working in this case?  Why am I hitting a GC wall at nowhere near the heap size?

**Comment by cantor-chron** [comment]
Did you try to decrease the default warmup iteration count, the iteration count, and the fork count? Maybe it's just that the algorithm is taking a long time.

```
  override def defaultConfig = Context(
    exec.minWarmupRuns -> 1,
    exec.maxWarmupRuns -> 1,
    exec.benchRuns -> 4,
    exec.independentSamples -> 1,
    verbose -> true
  )
```

This is taken from this example:

https://github.com/reactors-io/reactors/blob/master/reactors-common/jvm/src/bench/scala/io/reactors/common/concurrent/cache-trie-benches.scala

Another example if you're using it from Java (at the bottom of the page):

http://scalameter.github.io/home/gettingstarted/0.7/javausage/index.html

Also, if you add `verbose -> true`, you will see more detailed output about what ScalaMeter is doing.

**Comment by cantor-chron** [comment]
The default ScalaMeter benchmark configuration is doing GC before the measurement, to reduce the probability that the GC. This is done by the `Measurer` trait. The GC's that you see are possibly manually requested.

You can, of course, disable this by implementing your own `Measurer` class, and specifying your measurer in the concrete test (i.e. override the `measurer` method). This is relatively easy to do, but it sounds that in your case the test is taking long because the default warmup iteration count might have to be reduced.

**Comment by maxhillaert** [comment]
Yeah I test a table aggregation function on a million rows and introduce
bad concatenation of lists.  Memory blowing up and slow just like it was in
production.

When I tone it down to ten thousand rows maximum it will complete the test
run.  Obviously when I fix the bug it can finish the million rows fine

It’s just that I want to make sure if someone reintroduces the
concatenation bug , it fails the test rather than get stuck in a loop
trying to measure it.  I say ‘loop’, but of course it could just take a
very long time.

My concern is the masses of GC detections in the verbose log, without it
appearing to be able to move on. It sure feels like it’s stuck, but I could
be wrong.  I’ll try your suggestions and get back to you with some logs.
I’m keen on using this framework.

**Comment by cantor-chron** [comment]
> It’s just that I want to make sure if someone reintroduces the
concatenation bug , it fails the test rather than get stuck in a loop
trying to measure it.

One approach could be to implement a custom `Measurer` that starts a new thread to run the workload, and then `stop`s the thread after a timeout or exits the process.

**Comment by maxhillaert** [comment]
Yeah that would be good approach. Like a maximum test run time and fail.
This way we don’t hog our CI machines unnecessarily.

**Comment by maxhillaert** [comment]
So I've tried creating a warmer with a timeout:

    var i = 0
        while (i < maxwarmups) {

          setup()
          nogc = true
          val start = System.nanoTime

          if(maxRunningTime.nonEmpty){
            val ftr = Future(f(i))
            Await.result(ftr, maxRunningTime.get )
          }
          else{
            f(i)
          }

It blows up the whole test suite of course, so a bit blunt.   Any idea's on how to let it end the test run more gracefully and continue with next test?

**Comment by cantor-chron** [comment]
I would use `Await.ready` here, as it does not throw an exception.

But, I would not use futures at all, since `Await` does not cancel the computation in the future - i.e. your job will still be running. Starting a new thread, and force-stopping it with `Thread.stop` would be the best way to go if you want to kill the algorithm.

I think that you would then have to implement a custom `Reporter` that checks if a timeout occurred, and then override the `reporter` method in your test to decide whether the test failed or passed.
